# This file is supposed to perform an EDA on the vehicles_us dataset that contains information on diferent vehicles in the US. This task is supposed to serve as a first step towards building a web app analizing this data

# Libraries

In [1]:
import pandas as pd

# Helpful Functions

In [44]:
ds.shape

(51525, 13)

## Dealing with NaN

In [74]:
def how_many_nan(df, show_nan=False):
    """ This function will show us how many NaN are in each column and either show all the NaN values """
    
    # First we are gonna show all the nan if requested, otherwise show the nan count for each column
    if show_nan:
        return df[df.isnull().any(axis=1)]
    
    for i in df.keys():
        print(f'{i}: {df.shape[0] - df[i].count()}')
    
    print('\n\n\n')

        
    

# Loading the Data

In [2]:
ds = pd.read_csv('../vehicles_us.csv')
ds

price  model_year           model  condition  cylinders fuel  odometer  \
0       9400      2011.0          bmw x5       good        6.0  gas  145000.0   
1      25500         NaN      ford f-150       good        6.0  gas   88705.0   
2       5500      2013.0  hyundai sonata   like new        4.0  gas  110000.0   
3       1500      2003.0      ford f-150       fair        8.0  gas       NaN   
4      14900      2017.0    chrysler 200  excellent        4.0  gas   80903.0   
...      ...         ...             ...        ...        ...  ...       ...   
51520   9249      2013.0   nissan maxima   like new        6.0  gas   88136.0   
51521   2700      2002.0     honda civic    salvage        4.0  gas  181500.0   
51522   3950      2009.0  hyundai sonata  excellent        4.0  gas  128000.0   
51523   7455      2013.0  toyota corolla       good        4.0  gas  139573.0   
51524   6300      2014.0   nissan altima       good        4.0  gas       NaN   

      transmission    type paint_color  is_4wd date_posted  days_listed  
0        automatic     SUV         NaN     1.0  2018-06-23           19  
1        automatic  pickup       white     1.0  2018-10-19           50  
2        automatic   sedan         red     NaN  2019-02-07           79  
3        automatic  pickup         NaN     NaN  2019-03-22            9  
4        automatic   sedan       black     NaN  2019-04-02           28  
...            ...     ...         ...     ...         ...          ...  
51520    automatic   sedan       black     NaN  2018-10-03           37  
51521    automatic   sedan       white     NaN  2018-11-14           22  
51522    automatic   sedan        blue     NaN  2018-11-15           32  
51523    automatic   sedan       black     NaN  2018-07-02           71  
51524    automatic   sedan         NaN     NaN  2018-06-05           10  

[51525 rows x 13 columns]

# Taking a first look

In [3]:
ds.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51525 entries, 0 to 51524
Data columns (total 13 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   price         51525 non-null  int64  
 1   model_year    47906 non-null  float64
 2   model         51525 non-null  object 
 3   condition     51525 non-null  object 
 4   cylinders     46265 non-null  float64
 5   fuel          51525 non-null  object 
 6   odometer      43633 non-null  float64
 7   transmission  51525 non-null  object 
 8   type          51525 non-null  object 
 9   paint_color   42258 non-null  object 
 10  is_4wd        25572 non-null  float64
 11  date_posted   51525 non-null  object 
 12  days_listed   51525 non-null  int64  
dtypes: float64(4), int64(2), object(7)
memory usage: 5.1+ MB


In [4]:
ds.describe()

price    model_year     cylinders       odometer   is_4wd  \
count   51525.000000  47906.000000  46265.000000   43633.000000  25572.0   
mean    12132.464920   2009.750470      6.125235  115553.461738      1.0   
std     10040.803015      6.282065      1.660360   65094.611341      0.0   
min         1.000000   1908.000000      3.000000       0.000000      1.0   
25%      5000.000000   2006.000000      4.000000   70000.000000      1.0   
50%      9000.000000   2011.000000      6.000000  113000.000000      1.0   
75%     16839.000000   2014.000000      8.000000  155000.000000      1.0   
max    375000.000000   2019.000000     12.000000  990000.000000      1.0   

       days_listed  
count  51525.00000  
mean      39.55476  
std       28.20427  
min        0.00000  
25%       19.00000  
50%       33.00000  
75%       53.00000  
max      271.00000

Notice a few numbers that could lead to outliers but after a few strategies came to the conclusion that they are to few to make a difference

# Starting EDA

## Creating a copy of my dataframe since i dont wanna overwrite the original

In [78]:
df = ds.copy()

In the next few cells this copy doesnt come into action since im only looking at the dataframe. Later it will.

## Checking for duplicates

In [9]:
ds[ds.duplicated(keep='last')]

Empty DataFrame
Columns: [price, model_year, model, condition, cylinders, fuel, odometer, transmission, type, paint_color, is_4wd, date_posted, days_listed]
Index: []

## Looking at the nan values

In [22]:
ds[ds.isna().any(axis=1)]

price  model_year           model  condition  cylinders fuel  odometer  \
0       9400      2011.0          bmw x5       good        6.0  gas  145000.0   
1      25500         NaN      ford f-150       good        6.0  gas   88705.0   
2       5500      2013.0  hyundai sonata   like new        4.0  gas  110000.0   
3       1500      2003.0      ford f-150       fair        8.0  gas       NaN   
4      14900      2017.0    chrysler 200  excellent        4.0  gas   80903.0   
...      ...         ...             ...        ...        ...  ...       ...   
51520   9249      2013.0   nissan maxima   like new        6.0  gas   88136.0   
51521   2700      2002.0     honda civic    salvage        4.0  gas  181500.0   
51522   3950      2009.0  hyundai sonata  excellent        4.0  gas  128000.0   
51523   7455      2013.0  toyota corolla       good        4.0  gas  139573.0   
51524   6300      2014.0   nissan altima       good        4.0  gas       NaN   

      transmission    type paint_color  is_4wd date_posted  days_listed  
0        automatic     SUV         NaN     1.0  2018-06-23           19  
1        automatic  pickup       white     1.0  2018-10-19           50  
2        automatic   sedan         red     NaN  2019-02-07           79  
3        automatic  pickup         NaN     NaN  2019-03-22            9  
4        automatic   sedan       black     NaN  2019-04-02           28  
...            ...     ...         ...     ...         ...          ...  
51520    automatic   sedan       black     NaN  2018-10-03           37  
51521    automatic   sedan       white     NaN  2018-11-14           22  
51522    automatic   sedan        blue     NaN  2018-11-15           32  
51523    automatic   sedan       black     NaN  2018-07-02           71  
51524    automatic   sedan         NaN     NaN  2018-06-05           10  

[36673 rows x 13 columns]

In [76]:
how_many_nan(ds)

price: 0
model_year: 3619
model: 0
condition: 0
cylinders: 5260
fuel: 0
odometer: 7892
transmission: 0
type: 0
paint_color: 9267
is_4wd: 25953
date_posted: 0
days_listed: 0






In [75]:
how_many_nan(ds, True)

price  model_year           model  condition  cylinders fuel  odometer  \
0       9400      2011.0          bmw x5       good        6.0  gas  145000.0   
1      25500         NaN      ford f-150       good        6.0  gas   88705.0   
2       5500      2013.0  hyundai sonata   like new        4.0  gas  110000.0   
3       1500      2003.0      ford f-150       fair        8.0  gas       NaN   
4      14900      2017.0    chrysler 200  excellent        4.0  gas   80903.0   
...      ...         ...             ...        ...        ...  ...       ...   
51520   9249      2013.0   nissan maxima   like new        6.0  gas   88136.0   
51521   2700      2002.0     honda civic    salvage        4.0  gas  181500.0   
51522   3950      2009.0  hyundai sonata  excellent        4.0  gas  128000.0   
51523   7455      2013.0  toyota corolla       good        4.0  gas  139573.0   
51524   6300      2014.0   nissan altima       good        4.0  gas       NaN   

      transmission    type paint_color  is_4wd date_posted  days_listed  
0        automatic     SUV         NaN     1.0  2018-06-23           19  
1        automatic  pickup       white     1.0  2018-10-19           50  
2        automatic   sedan         red     NaN  2019-02-07           79  
3        automatic  pickup         NaN     NaN  2019-03-22            9  
4        automatic   sedan       black     NaN  2019-04-02           28  
...            ...     ...         ...     ...         ...          ...  
51520    automatic   sedan       black     NaN  2018-10-03           37  
51521    automatic   sedan       white     NaN  2018-11-14           22  
51522    automatic   sedan        blue     NaN  2018-11-15           32  
51523    automatic   sedan       black     NaN  2018-07-02           71  
51524    automatic   sedan         NaN     NaN  2018-06-05           10  

[36673 rows x 13 columns]

### Starting to work with model_year column

In [77]:
ds[ds['model_year'].isnull()]

price  model_year                       model  condition  cylinders  \
1      25500         NaN                  ford f-150       good        6.0   
20      6990         NaN             chevrolet tahoe  excellent        8.0   
65     12800         NaN                  ford f-150  excellent        6.0   
69      7800         NaN                  ford f-150   like new        8.0   
72      3650         NaN              subaru impreza  excellent        NaN   
...      ...         ...                         ...        ...        ...   
51464   8890         NaN        ford f250 super duty       good        8.0   
51465  34595         NaN                   gmc yukon  excellent        8.0   
51487   1495         NaN                  ford f-150       fair        6.0   
51488  21900         NaN  chevrolet silverado 3500hd  excellent        8.0   
51508   4950         NaN     chrysler town & country  excellent        6.0   

         fuel  odometer transmission      type paint_color  is_4wd  \
1         gas   88705.0    automatic    pickup       white     1.0   
20        gas  147485.0    automatic       SUV      silver     1.0   
65        gas  108500.0    automatic    pickup       white     NaN   
69        gas   97510.0    automatic     truck       white     1.0   
72        gas   74000.0    automatic     sedan        blue     1.0   
...       ...       ...          ...       ...         ...     ...   
51464     gas  135778.0    automatic     truck         red     1.0   
51465     gas   86456.0    automatic       SUV      silver     1.0   
51487     gas   15804.0       manual     truck       white     1.0   
51488  diesel  203000.0    automatic     truck       white     1.0   
51508     gas  150000.0    automatic  mini-van      silver     NaN   

      date_posted  days_listed  
1      2018-10-19           50  
20     2018-08-05           28  
65     2018-09-23           15  
69     2019-02-20           39  
72     2018-08-07           60  
...           ...          ...  
51464  2018-11-28           24  
51465  2019-04-18           13  
51487  2018-09-18           56  
51488  2018-12-27           37  
51508  2018-06-30           48  

[3619 rows x 13 columns]

I see model_year accounts for 10% of all the NaN. Im gonna try to use the values of other columns to fill those NaN, since simply using mean or median seems incorrect

In [80]:
# First starting to use the price column as an indicator to the year
df['model_year'] = df.groupby('price')['model_year'].apply(lambda x: x.fillna(x.mean())).reset_index(drop=True)

Checking if still any nan

In [87]:
df['model_year'].shape[0]-df['model_year'].count()

109

Only 109 left. Just gonna use the median on those since it seems more correct than the mean

In [119]:
df.describe()

price    model_year     cylinders       odometer   is_4wd  \
count   51525.000000  51525.000000  46265.000000   43633.000000  25572.0   
mean    12132.464920   2009.908722      6.125235  115553.461738      1.0   
std     10040.803015      5.686292      1.660360   65094.611341      0.0   
min         1.000000   1908.000000      3.000000       0.000000      1.0   
25%      5000.000000   2007.000000      4.000000   70000.000000      1.0   
50%      9000.000000   2011.000000      6.000000  113000.000000      1.0   
75%     16839.000000   2013.342466      8.000000  155000.000000      1.0   
max    375000.000000   2019.000000     12.000000  990000.000000      1.0   

       days_listed  
count  51525.00000  
mean      39.55476  
std       28.20427  
min        0.00000  
25%       19.00000  
50%       33.00000  
75%       53.00000  
max      271.00000

In [120]:
df['model_year'] = df['model_year'].fillna(df['model_year'].median())

Checking again for NaN

In [121]:
df['model_year'].shape[0]-df['model_year'].count()


0

### Moving on to the cylinders column

Gonna use the model_year and the model columns to fill the cylinders

In [123]:
df['cylinders'] = df.groupby(['model_year', 'model'])['cylinders'].apply(lambda x: x.fillna(x.median())).reset_index(drop=True)


c:\Users\Tirso\AppData\Local\Programs\Python\Python311\Lib\site-packages\numpy\lib\nanfunctions.py:1215: RuntimeWarning: Mean of empty slice
  return np.nanmean(a, axis, out=out, keepdims=keepdims)
c:\Users\Tirso\AppData\Local\Programs\Python\Python311\Lib\site-packages\numpy\lib\nanfunctions.py:1215: RuntimeWarning: Mean of empty slice
  return np.nanmean(a, axis, out=out, keepdims=keepdims)
c:\Users\Tirso\AppData\Local\Programs\Python\Python311\Lib\site-packages\numpy\lib\nanfunctions.py:1215: RuntimeWarning: Mean of empty slice
  return np.nanmean(a, axis, out=out, keepdims=keepdims)
c:\Users\Tirso\AppData\Local\Programs\Python\Python311\Lib\site-packages\numpy\lib\nanfunctions.py:1215: RuntimeWarning: Mean of empty slice
  return np.nanmean(a, axis, out=out, keepdims=keepdims)
c:\Users\Tirso\AppData\Local\Programs\Python\Python311\Lib\site-packages\numpy\lib\nanfunctions.py:1215: RuntimeWarning: Mean of empty slice
  return np.nanmean(a, axis, out=out, keepdims=keepdims)
c:\Users\T

In [125]:
df['cylinders'].shape[0]-df['cylinders'].count()

298